libgnucash/app-utils/gnc-option-date.cpp

```diff
@@ -0,0 +1,562 @@
+}
+const char*
+gnc_relative_date_description(RelativeDatePeriod per)
+{
+    if (per == RelativeDatePeriod::ABSOLUTE)
+        return nullptr;
+    return checked_reldate(per).m_description;
+}
+
+RelativeDatePeriod
+gnc_relative_date_from_storage_string(const char* str)
+{
+    auto per = std::find_if(reldates.begin(), reldates.end(),
+                         [str](auto rel) -> bool
+                         {
+                             return strcmp(str, rel.m_storage) == 0;
+                         });
+    return per != reldates.end() ? per->m_period : RelativeDatePeriod::ABSOLUTE;
+}
+
+static bool
+reldate_is_prev(RelativeDatePeriod per)
+{
+    auto rdate{checked_reldate(per)};
+    return per == RelativeDatePeriod::START_PREV_YEAR ||
+        per == RelativeDatePeriod::END_PREV_YEAR ||
+        per == RelativeDatePeriod::START_PREV_QUARTER ||
+        per == RelativeDatePeriod::END_PREV_QUARTER ||
+        per == RelativeDatePeriod::START_PREV_MONTH ||
+        per == RelativeDatePeriod::END_PREV_MONTH ||
+        rdate.m_type == LAST;
+}
+
+static bool
+reldate_is_next(RelativeDatePeriod per)
+{
+    auto rdate{checked_reldate(per)};
+    return per == RelativeDatePeriod::START_NEXT_YEAR ||
+        per == RelativeDatePeriod::END_NEXT_YEAR ||
+        per == RelativeDatePeriod::START_NEXT_QUARTER ||
+        per == RelativeDatePeriod::END_NEXT_QUARTER ||
+        per == RelativeDatePeriod::START_NEXT_MONTH ||
+        per == RelativeDatePeriod::END_NEXT_MONTH ||
+        rdate.m_type == NEXT;
+}
+
+static RelativeDateOffset
+reldate_offset(RelativeDatePeriod per)
+{
+    return checked_reldate(per).m_offset;
+}
+
+static constexpr int days_in_month[12]{31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
+
+static void
+normalize_tm(struct tm& now)
+{
+    auto tmp_mon = now.tm_mon + (now.tm_mon < 0 ? 12 :
+                                 now.tm_mon > 11 ? -12 : 0);
+    if (now.tm_mday < 1)
+    {
+        --now.tm_mon;
+        now.tm_mday += days_in_month[tmp_mon - 1];
+    }
+    else if (now.tm_mday > days_in_month[tmp_mon])
+    {
+        ++now.tm_mon;
+        now.tm_mday -= days_in_month[tmp_mon];
+    }
+    if (now.tm_mon < 0)
+    {
+        now.tm_mon += 12;
+        --now.tm_year;
+    }
+    else if (now.tm_mon > 11)
+    {
+        now.tm_mon -= 12;
+        ++now.tm_year;
+    }
+}
+
+static void
+set_day_and_time(struct tm& now, RelativeDateType type)
+{
+    if (type == RelativeDateType::START)
+    {
+        now.tm_hour = now.tm_min = now.tm_sec = 0;
+        now.tm_mday = 1;
+    }
+    else if (type == RelativeDateType::END)
+    {
+        now.tm_min = now.tm_sec = 59;
+        now.tm_hour = 23;
+        now.tm_mday = days_in_month[now.tm_mon];
+        // Check for Februrary in a leap year
+        if (int year = now.tm_year + 1900; now.tm_mon == 1 &&
+            year % 4 == 0 && (year % 100 != 0 || year % 400 == 0))
+            ++now.tm_mday;
+    }
+    // Do nothing for LAST and NEXT.
+};
```

Comment:
It took me a while to figure out what is happening here. While normalize_tm and set_day_and_time are internal functions it may be worth a comment to warn they are very tailored to the use case of converting our own relative date specs to absolute date/times. For example I was noticing normalize will only compensate tm_mon from -12 to -23 months or that leap year day compensation doesn't happen in normalize but in set_day_and_time. That's fine in itself but I was reading these functions initially as more generic due to their short names.

Reply:
Well, the short names (no gnc in front) and their `static` scope means to me that they're specific to this file which is pretty much just about GnuCash's RelativeDateTypes; there are more generally applicable versions in libgnucash/engine/gnc-datetime.cpp that I could have used but I wanted to avoid the expense of instantiating a GncDateTme and dealing with its exceptions.

Does renaming the functions to `normalize_reldate_tm` and `relate_set_day_and_time` help?

Reply:
I agree with your reasoning. No discussion there.

Here's my train of thought as an external reviewer when I first looked at this part of the code:
1. I saw the static scope too yet at the same time my first reflection was "why isn't this part of our GncDateTime" (which you have now motivated)
2.  next I noticed the implicit assumptions I mentioned and fought a bit with them in my mind. There was no mention of what would happen if the month part were further than one year off,  nor that calling code had to ensure this never happening. Or why leap year compensation only happens in the second helper,...
3.  and only then I saw `gnc_relative_date_to_time64` which actually uses these two functions. The math in there implicitly makes the inputs adhere to the assumptions so technically there's no problem in the code as it is now.

I mainly raised this as there are assumptions on input restraints in this bit of code that may not be obvious to all devs in all contexts. I did figure it out but it was non-obvious to **me** to start with. I had to pause for that.
In the future some dev may have ideas to alter or extend our reldate handling in ways that are no longer compatible with these assumptions but may fail to see that as it's non-obvious and hence break the code.

I don't want to make a fuss about this though so use your own judgement if you think an explicit mention of these implicit restraints in this specific context is needed.

I would certainly change the functions names as you propose.

Reply:
I can use `gnc_tm_set_day_start`, `gnc_tm_set_day_end` and `gnc_date_get_last_mday` in `reldate_set_day_and_time`. Does the same thing but hides the leap year check.

As for the implicit -12 to +23 constraint on `now.tm_mon` in `normalize_reldate_tm` I realize that it's conceivable that a day and month change could conceivably break that by calling ONE_YEAR_AHEAD from February 29, so I'll generalize it.

gnc-option-date also needs unit tests.